src/librustc_typeck/diagnostics.rs

```diff
@@ -1368,6 +1368,62 @@ struct Foo {
 ```
 "##,
 
+E0128: r##"
+Type parameter defaults can only use parameters that occur before them.*
+Erroneous code example:
+
+```
+pub struct Foo<T=U, U=()> {
+    field1: T,
+    filed2: U,
+}
+// error: type parameters with a default cannot use forward declared
+// identifiers
+```
+
+Since type parameters are evaluated in-order, you can fix this issue by doing:
+
+```
+pub struct Foo<U=(), T=U> {
+    field1: T,
+    filed2: U,
+}
+```
+
+Please also verify that a name clash hasn't been accidentally introduced, and
```

Comment:
Nit: name-clashes can't be introduced by reordering. Maybe "that this wasn't because of a name-clash".